### PR TITLE
Add VotingRequestFullValidationRule

### PR DESCRIPTION
**Pull request by quantumagi**
Merge #672 first

### CODE REVIEW

File: src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+            return Task.CompletedTask;
+        }
+
+        public void CheckTransaction(Transaction transaction)
+        {
+            if (transaction.IsCoinBase || transaction.IsCoinStake)
+                return;
+
+            // This will raise a consensus error if there is a voting request and it can't be decoded.
```
Comment (fassadlr) [inline src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs:54]: I dont think this comment is right here?

File: src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            JoinFederationRequest request = JoinFederationRequestBuilder.Deconstruct(transaction, this.encoder);
+            if (request == null)
+                return;
+
+            if (this.federationManager.IsMultisigMember(request.PubKey))
+            {
+                this.Logger.LogTrace("(-)[INVALID_MULTISIG_VOTING]");
```
Comment (fassadlr) [inline src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs:61]:
```suggestion
                this.Logger.LogTrace("(-)[MULTISIG_MEMBER_CANNOT_VOTE]");
```

File: src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                PoAConsensusErrors.VotingRequestInvalidMultisig.Throw();
```
Comment (fassadlr) [inline src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs:62]: Can you maybe rename this to `MultSigMemberCannotVote` adn the message within that class?

File: src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            {
+                this.Logger.LogTrace("(-)[INVALID_COLLATERAL_REQUIREMENT]");
+                PoAConsensusErrors.InvalidCollateralRequirement.Throw();
+            }
+
+            // Prohibit re-use of collateral addresses.
+            Script script = PayToPubkeyHashTemplate.Instance.GenerateScriptPubKey(request.CollateralMainchainAddress);
+            string collateralAddress = script.GetDestinationAddress(this.counterChainNetwork).ToString();
+            CollateralFederationMember owner = this.federationManager.CollateralAddressOwner(this.votingManager, VoteKey.AddFederationMember, collateralAddress);
+            if (owner != null && owner.PubKey != request.PubKey)
```
Comment (fassadlr) [inline src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs:78]: I think that the `owner.PubKey != request.PubKey` can actually be dropped? As the rule is that we abolsutely cannot have a duplicate address regardless of pubkey?

File: src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs (hunk 1 above)
Comment (quantumagi) [reply src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs:54]: `Deconstruct` either returns `null` if its not a voting request or calls `Decode` if it is. `Decode` in turn will raise consensus errors if a voting request is not properly constructed.

File: src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs (hunk 3 above)
Comment (quantumagi) [reply src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs:62]: Changed to:

`public static ConsensusError InvalidVotingOnMultiSig => new ConsensusError("invalid-voting-on-multisig", "invalid voting on multisig member");`

File: src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs (hunk 4 above)
Comment (quantumagi) [reply src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs:78]: It's just checking that a given collateral address is never associated with more than one public key. Removing that test would simply prevent the request initiator from wasting money if such a valid relationship already exists, and as such I agree.

File: src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs (hunk 2 above)
Comment (quantumagi) [reply src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs]: The new message is actually more misleading as it implies that a multisig member is attempting to vote. Instead it is the case that a multisig member cannot be voted on. I've changed this to `INVALID_VOTING_ON_MULTISIG_MEMBER`.

Comment (fassadlr) [reply src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs]: Ahhh OK yes that makes more sense now! 💯

**Review by fassadlr — APPROVED** [review]

**Pull request merged**